nighthawk/source/common/statistic_impl.cc

```diff
@@ -0,0 +1,216 @@
+  combined->streaming_stats_ = this->streaming_stats_->combine(*b.streaming_stats_);
+  return combined;
+}
+
+const int HdrStatistic::SIGNIFICANT_DIGITS = 4;
+
+HdrStatistic::HdrStatistic() : histogram_(nullptr) {
+  // Upper bound of 60 seconds (tracking in nanoseconds).
+  const uint64_t max_latency = 1000L * 1000 * 1000 * 60;
+
+  int status = hdr_init(1 /* min trackable value */, max_latency, HdrStatistic::SIGNIFICANT_DIGITS,
+                        &histogram_);
+  if (status != 0) {
+    ENVOY_LOG(error, "Failed to initialize HdrHistogram.");
+    histogram_ = nullptr;
+  }
+}
+
+// TODO(oschaaf): valgrind complains when a Histogram is created but never used.
+HdrStatistic::~HdrStatistic() {
+  if (histogram_ != nullptr) {
+    hdr_close(histogram_);
+    histogram_ = nullptr;
+  }
+}
+
+void HdrStatistic::addValue(int64_t value) {
+  if (histogram_ != nullptr) {
+    if (!hdr_record_value(histogram_, value)) {
+      ENVOY_LOG(warn, "Failed to record value into HdrHistogram.");
+    }
+  }
+}
+
+uint64_t HdrStatistic::count() const { return histogram_->total_count; }
+double HdrStatistic::mean() const { return hdr_mean(histogram_); }
+double HdrStatistic::variance() const {
+  return stdev() * stdev();
+  ;
+}
+double HdrStatistic::stdev() const {
+  // HdrHistogram_c's stdev actually gives us the population standard deviation.
+  // So we compute the sample standard deviation ourselves instead.
+  // TODO(oschaaf): this fixes some of the test expectations, but figure out if
+  // stdev or pstdev is preferrable. Looks like wrk2 uses pstdev which would produce
+  // (slightly) better numbers, though that probably isn't a reason for us to decice
+  // which one to use here. Switching to pstdev would get rid of having to do this
+  // ourselves.
+  if (histogram_ == nullptr) {
+    return 0;
+  }
+  double mean = hdr_mean(histogram_);
+  double geometric_dev_total = 0.0;
+
+  struct hdr_iter iter;
+  hdr_iter_init(&iter, histogram_);
+
+  while (hdr_iter_next(&iter)) {
+    if (0 != iter.count) {
+      double dev = (hdr_median_equivalent_value(histogram_, iter.value) * 1.0) - mean;
+      geometric_dev_total += (dev * dev) * iter.count;
+    }
+  }
+
+  return sqrt(geometric_dev_total / (histogram_->total_count - 1));
+}
+
+std::unique_ptr<Statistic> HdrStatistic::combine(const Statistic& statistic) {
+  auto combined = std::make_unique<HdrStatistic>();
+  const HdrStatistic& b = dynamic_cast<const HdrStatistic&>(statistic);
+
+  if (this->histogram_ == nullptr || b.histogram_ == nullptr) {
+    return combined;
+  }
+
+  int dropped;
+  dropped = hdr_add(combined->histogram_, this->histogram_);
+  dropped += hdr_add(combined->histogram_, b.histogram_);
+  if (dropped > 0) {
+    ENVOY_LOG(warn, "Combining HdrHistograms dropped values.");
+  }
+  return combined;
+}
+
+std::unique_ptr<HdrStatistic> HdrStatistic::getCorrected(const Frequency& frequency) {
```

Comment:
I'm still curious about the math and algorithm behind this correction for coordinated omission? Do you have references? Maybe add them as comments?

Reply:
I agree with removing this for now. The correction applied is based on the actual frequency vs the expected frequency; in the quick tests I ran with it I don't think it is of any use to us. It's probably better to track time spend blocked in closed-loop mode ourselves and report a second Statistic (or histogram) based on that somewhere in the future.